docs/decisions/0004-explicitly-declare-enterprise-for-role-assignments.rst

```diff
@@ -0,0 +1,74 @@
+Enterprise role assignments will record a customer UUID
+-------------------------------------------------------
+
+Status
+======
+
+Accepted - January 2021
+
+Context
+=======
+
+The ``SystemWideEnterpriseUserRoleAssignment`` model is used to grant role-based permissions
+to enterprise users.  These permissions generally control the users' access to data
+and ability to perform certain functions within an enterprise customer.  This assignment model does `not` currently
+refer to any specific customer - the "context" to which a role is applied is inferred from
+a user (an ``auth.User`` record) having been "linked" to a customer via the ``EnterpriseCustomerUser`` model.
+
+However, a single ``auth.User`` may be linked to multiple ``EnterpriseCustomers``, and if that user
+is assigned a role (such as ``enterprise_admin``), it is inferred that the role assignment applies
+to **every** customer to which the user is linked.  This is undesirable; we generally do not want an admin
+of `Customer A` to also have admin permissions for `Customer B` (to which the user may be linked as a mere ``learner``).
+Instead, the ``SystemWideEnterpriseUserRoleAssignment`` model should explicitly record *which* customer
+the assigned user has the given role in.
+
+Decisions (interspersed with Consequences)
+==========================================
+
+We'll add an ``enterprise_customer`` field to the assignment model
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* It will be an optional field.  It should be a foreign key on ``EnterpriseCustomer.uuid``.
+* Via an edx-rbac change, we will also add a boolean ``applies_to_all_contexts`` field,
+  which is a wildcard that, if true, indicates the use has the role assigned for every customer.
+  It should default to ``false``.
+* The semantics of ``applies_to_all_contexts`` imply that ``enterprise_customer`` will sometimes be null.
+  For example, a user assigned the ``enterprise_openedx_operator`` role will have ``applies_to_all_contexts``
```

Comment:
If something sets the enterprise_customer field, will it also set the `applies_to_all_contexts` field to false?

Reply:
Good question!  I'm going to say no (for now), but maybe we could add some messaging in the Django Admin form around this.

Reply:
ok yeah that may be helpful, but yeah just wondering of a case if the enterprise_customer is set and the applies* field is also true .. guess the latter will win?